File: apps/python/src/algorithm/sort.py
"""第6章 ソートアルゴリズム"""


def bubble_sort(a: list) -> None:
    """バブルソート（in-place）

    隣接する要素を比較・交換し、最大値を末尾へ浮き上がらせる。
    最適化: 交換が発生しなければ終了。

    計算量: O(n^2) 最悪, O(n) 最良（整列済み）

    >>> a = [6, 4, 3, 7, 1, 9, 8]
    >>> bubble_sort(a)
    >>> a
    [1, 3, 4, 6, 7, 8, 9]
    """
    n = len(a)
    for i in range(n - 1):
        swapped = False
        for j in range(n - 1, i, -1):
            if a[j - 1] > a[j]:
                a[j - 1], a[j] = a[j], a[j - 1]
                swapped = True
        if not swapped:
            break


def selection_sort(a: list) -> None:
    """選択ソート（in-place）

    未整列部分の最小値を探し、先頭と交換する。

    計算量: O(n^2)

    >>> a = [6, 4, 3, 7, 1, 9, 8]
    >>> selection_sort(a)
    >>> a
    [1, 3, 4, 6, 7, 8, 9]
    """
    n = len(a)
    for i in range(n - 1):
        min_idx = i
        for j in range(i + 1, n):
            if a[j] < a[min_idx]:
                min_idx = j
        if min_idx != i:
            a[i], a[min_idx] = a[min_idx], a[i]


def insertion_sort(a: list) -> None:
    """挿入ソート（in-place）

    未整列部分の先頭要素を、整列済み部分の適切な位置に挿入する。

    計算量: O(n^2) 最悪, O(n) 最良（整列済み）

    >>> a = [6, 4, 3, 7, 1, 9, 8]
    >>> insertion_sort(a)
    >>> a
    [1, 3, 4, 6, 7, 8, 9]
    """
    n = len(a)
    for i in range(1, n):
        key = a[i]
        j = i - 1
        while j >= 0 and a[j] > key:
            a[j + 1] = a[j]
            j -= 1
        a[j + 1] = key


def shell_sort(a: list) -> None:
    """シェルソート（in-place）

    挿入ソートの改良。間隔（gap）を縮小しながら複数回の挿入ソートを行う。
    Knuth 数列（1, 4, 13, 40, ...）を使用。

    計算量: O(n^(3/2)) ～ O(n log^2 n)

    >>> a = [6, 4, 3, 7, 1, 9, 8]
    >>> shell_sort(a)
    >>> a
    [1, 3, 4, 6, 7, 8, 9]
    """
    n = len(a)
    # Knuth 数列で最大 gap を決定
    gap = 1
    while gap * 3 + 1 < n:
        gap = gap * 3 + 1

    while gap > 0:
        for i in range(gap, n):
            key = a[i]
            j = i - gap
            while j >= 0 and a[j] > key:
                a[j + gap] = a[j]
                j -= gap
            a[j + gap] = key
        gap //= 3


def quick_sort(a: list, left: int = 0, right: int | None = None) -> None:
    """クイックソート（in-place）

    ピボットを基準に配列を分割し、再帰的にソートする。

    計算量: O(n log n) 平均, O(n^2) 最悪

    >>> a = [6, 4, 3, 7, 1, 9, 8]
    >>> quick_sort(a)
    >>> a
    [1, 3, 4, 6, 7, 8, 9]
    """
    if right is None:
        right = len(a) - 1

    if left >= right:
        return

    pivot = a[(left + right) // 2]
    i, j = left, right

    while i <= j:
        while a[i] < pivot:
            i += 1
        while a[j] > pivot:
            j -= 1
        if i <= j:
            a[i], a[j] = a[j], a[i]
            i += 1
            j -= 1

    quick_sort(a, left, j)
    quick_sort(a, i, right)


def merge_sort(a: list) -> list:
    """マージソート（新しいリストを返す）

    配列を半分に分割し、再帰的にソートして結合する。

    計算量: O(n log n)

    >>> merge_sort([6, 4, 3, 7, 1, 9, 8])
    [1, 3, 4, 6, 7, 8, 9]
    """
    if len(a) <= 1:
        return a[:]

    mid = len(a) // 2
    left = merge_sort(a[:mid])
    right = merge_sort(a[mid:])

    return _merge(left, right)


def _merge(left: list, right: list) -> list:
    """2 つの整列済みリストをマージ"""
    result = []
    i = j = 0
    while i < len(left) and j < len(right):
        if left[i] <= right[j]:
            result.append(left[i])
            i += 1
        else:
            result.append(right[j])
            j += 1
    result.extend(left[i:])
    result.extend(right[j:])
    return result

File: apps/ruby/lib/algorithm/sort.rb
# frozen_string_literal: true

# 第6章 ソートアルゴリズム

module Algorithm
  # バブルソート（in-place）
  def self.bubble_sort(a)
    n = a.length
    (n - 1).times do |i|
      swapped = false
      (n - 1).downto(i + 1) do |j|
        if a[j - 1] > a[j]
          a[j - 1], a[j] = a[j], a[j - 1]
          swapped = true
        end
      end
      break unless swapped
    end
  end

  # 選択ソート（in-place）
  def self.selection_sort(a)
    n = a.length
    (n - 1).times do |i|
      min_idx = i
      (i + 1...n).each do |j|
        min_idx = j if a[j] < a[min_idx]
      end
      a[i], a[min_idx] = a[min_idx], a[i] if min_idx != i
    end
  end

  # 挿入ソート（in-place）
  def self.insertion_sort(a)
    n = a.length
    (1...n).each do |i|
      key = a[i]
      j = i - 1
      while j >= 0 && a[j] > key
        a[j + 1] = a[j]
        j -= 1
      end
      a[j + 1] = key
    end
  end

  # シェルソート（in-place, Knuth 数列）
  def self.shell_sort(a)
    n = a.length
    gap = 1
    gap = gap * 3 + 1 while gap * 3 + 1 < n

    while gap > 0
      (gap...n).each do |i|
        key = a[i]
        j = i - gap
        while j >= 0 && a[j] > key
          a[j + gap] = a[j]
          j -= gap
        end
        a[j + gap] = key
      end
      gap /= 3
    end
  end

  # クイックソート（in-place）
  def self.quick_sort(a, left = 0, right = nil)
    right ||= a.length - 1
    return if left >= right

    pivot = a[(left + right) / 2]
    i = left
    j = right

    while i <= j
      i += 1 while a[i] < pivot
      j -= 1 while a[j] > pivot
      if i <= j
        a[i], a[j] = a[j], a[i]
        i += 1
        j -= 1
      end
    end

    quick_sort(a, left, j)
    quick_sort(a, i, right)
  end

  # マージソート（新しい配列を返す）
  def self.merge_sort(a)
    return a.dup if a.length <= 1

    mid = a.length / 2
    left = merge_sort(a[0...mid])
    right = merge_sort(a[mid..])
    _merge(left, right)
  end

  def self._merge(left, right)
    result = []
    i = j = 0
    while i < left.length && j < right.length
      if left[i] <= right[j]
        result << left[i]
        i += 1
      else
        result << right[j]
        j += 1
      end
    end
    result.concat(left[i..])
    result.concat(right[j..])
    result
  end
  private_class_method :_merge

  # ヒープソート（in-place）
  def self.heap_sort(a)
    n = a.length

    down_heap = lambda do |arr, left, right|
      temp = arr[left]
      parent = left
      while parent < (right + 1) / 2
        cl = parent * 2 + 1
        cr = cl + 1
        child = (cr <= right && arr[cr] > arr[cl]) ? cr : cl
        break if temp >= arr[child]

        arr[parent] = arr[child]
        parent = child
      end
      arr[parent] = temp
    end

    ((n - 1) / 2).downto(0) { |i| down_heap.call(a, i, n - 1) }
    (n - 1).downto(1) do |i|
      a[0], a[i] = a[i], a[0]
      down_heap.call(a, 0, i - 1)
    end
  end

  # 度数ソート（計数ソート）— 新しい配列を返す
  def self.counting_sort(a)
    return [] if a.empty?

    max_val = a.max
    freq = Array.new(max_val + 1, 0)
    a.each { |x| freq[x] += 1 }
    (1...freq.length).each { |i| freq[i] += freq[i - 1] }

    result = Array.new(a.length, 0)
    a.reverse_each do |x|
      freq[x] -= 1
      result[freq[x]] = x
    end
    result
  end
end
